### ISSUE TITLE
Inconsistent error message when attempting SSA with the wrong uid

### ISSUE THREAD

**Issue by mdbooth** [kubernetes/kubernetes#127625]
### What happened?

I am specifying uid in a Server-Side Apply transaction to ensure that the object I am patching is the object I am expecting to patch, and to ensure I am not accidentally creating a new object. When specifying the wrong uid, the error message is:
```
The ConfigMap "foo" is invalid: metadata.uid: Invalid value: "014dbd28-1f25-42cb-a79c-955b98409d02": field is immutable
```

This is an error message about attempting to modify a value. Compare this to the error message from specifying an out-of-date resourceVersion:
```
error: Operation cannot be fulfilled on configmaps "foo": the object has been modified; please apply your changes to the latest version and try again
```

This is explicit about what has occurred.

### What did you expect to happen?

An error message which is clearer that an attempt was made to patch the wrong object. Maybe:
```
error: Operation cannot be fulfilled on configmaps "foo": patch specifies a different uid
```

### How can we reproduce it (as minimally and precisely as possible)?

`cm-test.yaml`:
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: foo
  namespace: default
```

```
$ kubectl apply -f cm-test.yaml --server-side
```

Update `cm-test.yaml` with a uid different to whatever was generated:
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: foo
  namespace: default
  uid: e5a88576-fcda-43ae-88af-4ca2a6087a16
```

```
$ kubectl apply -f cm-test.yaml --server-side
The ConfigMap "foo" is invalid: metadata.uid: Invalid value: "e5a88576-fcda-43ae-88af-4ca2a6087a16": field is immutable
```

### Anything else we need to know?

There may be additional context in https://kubernetes.slack.com/archives/C0EG7JC6T/p1727274176568809

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: v1.28.4
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.30.0
WARNING: version difference between client (1.28) and server (1.30) exceeds the supported minor version skew of +/-1
```

</details>

### Cloud provider

<details>
N/A
</details>

### OS version

<details>
N/A
</details>

### Install tools

<details>
N/A
</details>

### Container runtime (CRI) and version (if applicable)

<details>
N/A
</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>
N/A
</details>

**Comment by sftim** [comment]
Kubernetes works as specified. We can improve the error message; to me that's a small feature request.

/remove-kind bug
/kind feature

**Comment by pranav-pandey0804** [comment]
The current error message makes it sound like you're trying to modify the UID, which isn't really the issue here.

**Comment by pranav-pandey0804** [comment]
A clearer message that highlights the mismatch between the `uid` specified and the one on the object would avoid confusion and make it easier to diagnose.

**Comment by pranav-pandey0804** [comment]
yes @sftim, this is more of a small feature request than a bug, as the error message should better align with the intent of the operation. Improving the clarity of this message, similar to the one used for `resourceVersion`, would enhance the user experience without altering the underlying functionality.

**Comment by pranav-pandey0804** [comment]
I've assigned this issue to myself.
Current error handling treats the UID mismatch as an attempt to modify an immutable field, which isn't clear. 
I'll be working on updating the error message to reflect a UID mismatch more accurately, ensuring it provides better clarity for users.
